hotel/models/hotel_folio.py

```diff
@@ -0,0 +1,858 @@
+        "Check In",
+        required=True,
+        readonly=True,
+        states={"draft": [("readonly", False)]},
+        default=_get_checkin_date,
+    )
+    checkout_date = fields.Datetime(
+        "Check Out",
+        required=True,
+        readonly=True,
+        states={"draft": [("readonly", False)]},
+        default=_get_checkout_date,
+    )
+    room_lines = fields.One2many(
+        "hotel.folio.line",
+        "folio_id",
+        readonly=True,
+        states={"draft": [("readonly", False)], "sent": [("readonly", False)]},
+        help="Hotel room reservation detail.",
+    )
+    service_lines = fields.One2many(
+        "hotel.service.line",
+        "folio_id",
+        readonly=True,
+        states={"draft": [("readonly", False)], "sent": [("readonly", False)]},
+        help="Hotel services details provided to"
+        "Customer and it will included in "
+        "the main Invoice.",
+    )
+    hotel_policy = fields.Selection(
+        [
+            ("prepaid", "On Booking"),
+            ("manual", "On Check In"),
+            ("picking", "On Checkout"),
+        ],
+        "Hotel Policy",
+        default="manual",
+        help="Hotel policy for payment that "
+        "either the guest has to payment at "
+        "booking time or check-in "
+        "check-out time.",
+    )
+    duration = fields.Float(
+        "Duration in Days",
+        help="Number of days which will automatically "
+        "count from the check-in and check-out date. ",
+    )
+    hotel_invoice_id = fields.Many2one("account.move", "Invoice", copy=False)
+    duration_dummy = fields.Float("Duration Dummy")
+
+    @api.constrains("room_lines")
+    def folio_room_lines(self):
+        """
+        This method is used to validate the room_lines.
+        ------------------------------------------------
+        @param self: object pointer
+        @return: raise warning depending on the validation
+        """
+        # folio_rooms = []
+        # for room in self[0].room_lines:
+        #     if room.product_id.id in folio_rooms:
+        #         raise ValidationError(_("You Cannot Take Same Room Twice"))
+        #     folio_rooms.append(room.product_id.id)
+
+        for rec in self:
+            for room_no in rec.room_lines.mapped("product_id"):
+                lines = rec.room_lines.search(
+                    [
+                        ("product_id", "=", room_no.id),
+                        ("folio_id", "=", rec.id),
+                    ]
+                )
+                for line in lines:
+                    record = lines.search(
+                        [
+                            ("id", "!=", line.id),
+                            ("id", "in", lines.ids),
+                            ("checkin_date", ">=", line.checkin_date),
+                            ("checkout_date", "<=", line.checkout_date),
+                            ("product_id", "=", room_no.id),
+                        ]
+                    )
```

Comment:
you can use the collections class for to check duplicate
room_ids = rec.room_lines.mapped("product_id").ids
duplicate = [item for item,count in collections.Counter(room_ids).items() if count > 1]

Reply:
Well! In this we need products with date filter.which is not possible by using collections.Counter. But code optimisation is possible by merging both for loop.